### PR TITLE
fix(api): get revert message in eth_call

### PR DESCRIPTION
**Pull request by hunshenshi**
# Description
Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #(issue)
#3821 

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/web3server.go [hunk 1]
```diff
@@ -391,6 +395,21 @@ func (svr *web3Handler) call(in *gjson.Result) (interface{}, error) {
...
+		arguments := abi.Arguments{{Type: strtyp}}
+		args, err := arguments.Unpack(result[4:])
+		if err != nil {
+			log.Logger("api").Error("web3server",
+				zap.String("requestParams", "ret abi unpack"),
+				zap.Error(err))
+			return "0x" + ret, nil
+		}
+		return "0x" + ret, status.Error(codes.InvalidArgument, fmt.Sprintf("execution reverted: %s", args[0].(string)))
+	}
```
Comment (envestcc) [inline api/web3server.go:411]: we already have `ExecutionRevertMsg` field in `iotextypes.Receipt` at returned by `ReadContract`, please try to use it first if possible.

Comment (envestcc) [reply api/web3server.go:411]: Your code seems to attempt manual parsing of EVM return codes, which is not recommended. It may pose obstacles for future EVM upgrades. 
If `ExecutionRevertMsg` not work, maybe you can refer to this PR  in Ethereum, using `abi.UnpackRevert` for parsing purpose:
https://github.com/ethereum/go-ethereum/pull/21083/files#diff-c426ecd2f7d247753b9ea8c1cc003f21fa412661c1f967d203d4edf8163da344R912-R914

**Comment by millken** [comment]
your code has broken ci test, and need a test to verify your changes. looks like a wrong fix

File: api/web3server.go [hunk 2]
```diff
@@ -387,10 +389,15 @@ func (svr *web3Handler) call(in *gjson.Result) (interface{}, error) {
...
 	}
 	exec, _ := action.NewExecution(to, 0, value, gasLimit, big.NewInt(0), data)
-	ret, _, err := svr.coreService.ReadContract(context.Background(), callerAddr, exec)
+	ret, receipt, err := svr.coreService.ReadContract(context.Background(), callerAddr, exec)
 	if err != nil {
 		return nil, err
 	}
+
+	if len(receipt.ExecutionRevertMsg) > 0 {
+		return "0x" + ret, status.Error(codes.InvalidArgument, receipt.ExecutionRevertMsg)
```
Comment (millken) [inline api/web3server.go:398]: please verify that you get the expected API response
```
{
    "jsonrpc":"2.0",
    "id":1,
    "error":{
         "code":3,
         "message":"execution reverted: require test",
         "data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000c7265717569726520746573740000000000000000000000000000000000000000"
    }
}
```

Comment (hunshenshi) [reply api/web3server.go:398]: > please verify that you get the expected API response
> 
> ```
> {
>     "jsonrpc":"2.0",
>     "id":1,
>     "error":{
>          "code":3,
>          "message":"execution reverted: require test",
>          "data":"0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000c7265717569726520746573740000000000000000000000000000000000000000"
>     }
> }
> ```
done, pls review

File: action/protocol/execution/evm/evm.go [hunk 3]
```diff
@@ -292,7 +293,7 @@ func ExecuteContract(
 		data := retval[4:]
 		msgLength := byteutil.BytesToUint64BigEndian(data[56:64])
 		revertMsg := string(data[64 : 64+msgLength])
-		receipt.SetExecutionRevertMsg(revertMsg)
+		receipt.SetExecutionRevertMsg(fmt.Sprintf("execution reverted: %s", revertMsg))
```
Comment (envestcc) [inline action/protocol/execution/evm/evm.go:296]: It's better to insert "execution reverted:" at api/web3server.go instead of evm.go, leading to much unnecessary changes

File: api/web3server.go [hunk 4]
```diff
@@ -21,6 +21,8 @@ import (
 	"github.com/tidwall/gjson"
 	"go.opentelemetry.io/otel/attribute"
 	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
```
Comment (envestcc) [inline api/web3server.go:25]: unused pkg

Comment (hunshenshi) [reply api/web3server.go:25]: used here `status.Error(codes.InvalidArgument, receipt.ExecutionRevertMsg)`

**Review by envestcc — APPROVED** [review]

File: action/protocol/execution/evm/evm.go (hunk 3 above)
Comment (Liuhaai) [reply action/protocol/execution/evm/evm.go:296]: don't change it in EVM, which will cause a HF

File: api/web3server.go (hunk 2 above)
Comment (Liuhaai) [reply api/web3server.go:398]: `status` lib is used in grpc, not in http

**Comment by millken** [comment]
PR #3920 has included fix for Issue #3821, closed.
